WindowManager/app/src/main/java/com/example/windowmanagersample/DisplayFeaturesActivity.kt

```diff
@@ -19,59 +19,66 @@ package com.example.windowmanagersample
 
 /** Demo activity that shows all display features and current device state on the screen. */
-class DisplayFeaturesActivity : BaseSampleActivity() {
+class DisplayFeaturesActivity : AppCompatActivity() {
 
-    private lateinit var windowManager: WindowManager
     private val stateLog: StringBuilder = StringBuilder()
 
     private val displayFeatureViews = ArrayList<View>()
 
-    // Store most recent values for the device state and window layout
-    private val stateContainer = StateContainer()
     private lateinit var binding: ActivityDisplayFeaturesBinding
+    private lateinit var windowInfoRepo: WindowInfoRepo
 
+    @ExperimentalCoroutinesApi
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
+
         binding = ActivityDisplayFeaturesBinding.inflate(layoutInflater)
         setContentView(binding.root)
 
-        windowManager = getTestBackend()?.let { backend ->
-            binding.deviceStateToggleButton.visibility = View.VISIBLE
-            binding.deviceStateToggleButton.setOnClickListener {
-                if (backend is MidScreenFoldBackend) {
-                    backend.toggleDeviceHalfOpenedState(this)
-                }
+        windowInfoRepo = windowInfoRepository()
+
+        // Create a new coroutine since repeatOnLifecycle is a suspend function
+        lifecycleScope.launch {
+            // The block passed to repeatOnLifecycle is executed when the lifecycle
+            // is at least STARTED and is cancelled when the lifecycle is STOPPED.
+            // It automatically restarts the block when the lifecycle is STARTED again.
+            lifecycle.repeatOnLifecycle(Lifecycle.State.STARTED) {
+                // Safely collect from windowInfoRepo when the lifecycle is STARTED
+                // and stops collection when the lifecycle is STOPPED
+                windowInfoRepo.windowLayoutInfo()
+                    // Throttle first event 10ms to allow the UI to pickup the posture
+                    .throttleFirst(10)
+                    .collect { newLayoutInfo ->
+                        // New posture information
+                        updateStateLog(newLayoutInfo)
+                        updateCurrentState(newLayoutInfo)
+                    }
             }
-            WindowManager(this, backend)
-        } ?: WindowManager(this)
+        }
 
         stateLog.clear()
         stateLog.append(getString(R.string.state_update_log)).append("\n")
     }
 
-    override fun onStart() {
-        super.onStart()
-        windowManager.registerLayoutChangeCallback(mainThreadExecutor, stateContainer)
-    }
-
-    override fun onStop() {
-        super.onStop()
-        windowManager.unregisterLayoutChangeCallback(stateContainer)
-    }
-
     /** Updates the device state and display feature positions. */
-    internal fun updateCurrentState(layoutInfo: WindowLayoutInfo?) {
+    private fun updateCurrentState(layoutInfo: WindowLayoutInfo?) {
```

Comment:
Can layoutInfo be non-nullable now?

Reply:
Thanks for the comment. I forgot to update the parameter when converting from the callback flow.